File: Core-DSA-Optimized/Graphs/A-Star-Pathfinding/a_star.py
import heapq

class AStar:
    def __init__(self, grid):
        self.grid = grid
        self.rows = len(grid)
        self.cols = len(grid[0])

    def heuristic(self, a, b):
        # Manhattan distance on a square grid
        return abs(a[0] - b[0]) + abs(a[1] - b[1])

    def find_path(self, start, goal):
        # Priority Queue stores: (f_cost, current_node)
        open_set = []
        heapq.heappush(open_set, (0, start))
        
        came_from = {}
        
        # g_score: Cost from start to node
        g_score = {start: 0}
        
        # f_score: g_score + heuristic
        f_score = {start: self.heuristic(start, goal)}

        while open_set:
            current_f, current = heapq.heappop(open_set)

            if current == goal:
                path = []
                while current in came_from:
                    path.append(current)
                    current = came_from[current]
                path.append(start)
                return path[::-1] # Reverse to get Start -> Goal

            # Check 4 adjacent neighbors (Up, Down, Left, Right)
            neighbors = [(0, 1), (1, 0), (0, -1), (-1, 0)]
            for dx, dy in neighbors:
                neighbor = (current[0] + dx, current[1] + dy)
                
                # Check bounds and walls (0 = free space, 1 = wall)
                if 0 <= neighbor[0] < self.rows and 0 <= neighbor[1] < self.cols:
                    if self.grid[neighbor[0]][neighbor[1]] == 1:
                        continue

                    tentative_g_score = g_score[current] + 1

                    if neighbor not in g_score or tentative_g_score < g_score[neighbor]:
                        came_from[neighbor] = current
                        g_score[neighbor] = tentative_g_score
                        f_score[neighbor] = tentative_g_score + self.heuristic(neighbor, goal)
                        
                        # Only push if we found a strictly better path
                        heapq.heappush(open_set, (f_score[neighbor], neighbor))

        return [] # No path found

# --- CI/CD Automated Test ---
if __name__ == '__main__':
    # 0 = Path, 1 = Wall
    maze = [
        [0, 1, 0, 0, 0],
        [0, 1, 0, 1, 0],
        [0, 0, 0, 1, 0],
        [0, 1, 1, 1, 0],
        [0, 0, 0, 0, 0]
    ]
    
    astar = AStar(maze)
    start_node = (0, 0)
    end_node = (4, 4)
    
    path = astar.find_path(start_node, end_node)
    
    # Path should successfully navigate around the '1's
    assert len(path) > 0, "A* failed to find a path!"
    assert path[0] == start_node, "Path doesn't begin at start."
    assert path[-1] == end_node, "Path doesn't arrive at goal."
    
    print(f"Python A* Pathfinding Test Passed! Path length: {len(path)}")

File: Core-DSA-Optimized/Graphs/A-Star-Pathfinding/aStar.ts
type Node = { r: number, c: number };

export class AStar {
    private grid: number[][];
    private rows: number;
    private cols: number;

    constructor(grid: number[][]) {
        this.grid = grid;
        this.rows = grid.length;
        this.cols = grid[0].length;
    }

    private heuristic(a: Node, b: Node): number {
        return Math.abs(a.r - b.r) + Math.abs(a.c - b.c); // Manhattan
    }

    private hash(node: Node): string {
        return `${node.r},${node.c}`;
    }

    public findPath(start: Node, goal: Node): Node[] {
        // Priority Queue (Sorted dynamically for simplicity in single-file)
        let openSet: { f: number, node: Node }[] = [{ f: 0, node: start }];
        const cameFrom = new Map<string, Node>();
        
        const gScore = new Map<string, number>();
        gScore.set(this.hash(start), 0);

        while (openSet.length > 0) {
            openSet.sort((a, b) => a.f - b.f);
            const current = openSet.shift()!.node;

            if (current.r === goal.r && current.c === goal.c) {
                const path = [current];
                let currHash = this.hash(current);
                while (cameFrom.has(currHash)) {
                    const prev = cameFrom.get(currHash)!;
                    path.push(prev);
                    currHash = this.hash(prev);
                }
                return path.reverse();
            }

            const directions = [[0, 1], [1, 0], [0, -1], [-1, 0]];
            for (const [dr, dc] of directions) {
                const neighbor = { r: current.r + dr, c: current.c + dc };
                
                if (neighbor.r >= 0 && neighbor.r < this.rows && neighbor.c >= 0 && neighbor.c < this.cols) {
                    if (this.grid[neighbor.r][neighbor.c] === 1) continue;

                    const tentativeG = gScore.get(this.hash(current))! + 1;
                    const neighborHash = this.hash(neighbor);

                    if (!gScore.has(neighborHash) || tentativeG < gScore.get(neighborHash)!) {
                        cameFrom.set(neighborHash, current);
                        gScore.set(neighborHash, tentativeG);
                        const fScore = tentativeG + this.heuristic(neighbor, goal);
                        openSet.push({ f: fScore, node: neighbor });
                    }
                }
            }
        }
        return [];
    }
}

// --- CI/CD Automated Test ---
const maze = [
    [0, 1, 0, 0, 0],
    [0, 1, 0, 1, 0],
    [0, 0, 0, 1, 0],
    [0, 1, 1, 1, 0],
    [0, 0, 0, 0, 0]
];

const astar = new AStar(maze);
const path = astar.findPath({r: 0, c: 0}, {r: 4, c: 4});

if (path.length > 0 && path[0].r === 0 && path[path.length - 1].r === 4) {
    console.log(`TypeScript A* Pathfinding Test Passed! Path Length: ${path.length}`);
} else {
    process.exit(1);
}